### PR TITLE
[FIXED JENKINS-41272] Expose build causes from currentBuild

### PR DESCRIPTION
**Pull request by mpavlov**
[JENKINS-41272](https://issues.jenkins-ci.org/browse/JENKINS-41272)
cc @jglick

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/workflow/support/steps/build/RunWrapper.java [hunk 1]
```diff
@@ -240,4 +242,21 @@ public String getId() throws AbortException {
         }
     }
 
+    @Whitelisted
+    public List<Cause> getCauses() throws AbortException {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/support/steps/build/RunWrapper.java]: Only useful insofar as the `Cause` implementations themselves have whitelisted getters.

File: src/main/java/org/jenkinsci/plugins/workflow/support/steps/build/RunWrapper.java [hunk 2]
```diff
@@ -240,4 +242,21 @@ public String getId() throws AbortException {
...
+        return build().getCauses();
+    }
+
+    @Whitelisted
+    public List<String> getCauseDescriptions() throws AbortException {
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/workflow/support/steps/build/RunWrapper.java]: Would rather not expose this as it encourages scripts to rely on inherently unreliable localized messages.

**Comment by mpavlov** [comment]
@jglick I see the point in not enabling functionality that's easily misused.

Would it still be beneficial to solve the specific problem (JENKINS-41272) separately from the general `DescribableModel`-like POJO uninstantiation system?

For example, maybe something like this could greatly help the majority of Cause-based use cases?
```
    @Whitelisted
    public boolean getIsStartedByUser() {
        return build().getCause(Cause.UserIdCause.class) != null;
    }

    @Whitelisted
    public boolean getIsStartedByTimer() {
        return build().getCause(TimerTrigger.TimerTriggerCause.class) != null;
    }
```

**Comment by gabloe** [comment]
> For example, maybe something like this could greatly help the majority of Cause-based use cases?
It would be nice if the cause was an upstream project to be able to get the upstream project name as well.

**Comment by mpavlov** [comment]
@jglick I've removed the reliance on localized messages. Can you take another look?
@gabloe I've added `getUpstreamProjectName()`
cc: @dtschan

**Comment by smokris** [comment]
It would be helpful to also have a `getStartedByUserId()` method, so we can customize the build for the user who initiated it.

I don't have write access to this PR, but I think it would be something like this:

```groovy
@Whitelisted
public String getStartedByUserId() throws Exception {
    Cause.UserIdCause cause = build().getCause(Cause.UserIdCause.class);
    return cause == null ? null : cause.getUserId();
}
```

**Comment by mpavlov** [comment]
Based on the comments, the proposed change is not a good direction for a long-term solution. And it appears there is not enough interest to use it as an interim fix either. Closing.

**Pull request closed without merging**